sqlx-core/src/postgres/listen.rs

```diff
@@ -0,0 +1,216 @@
+
+impl<C> PgListener<C> {
+    /// Construct a new instance.
+    pub(self) fn new(conn: C) -> Self {
+        Self(conn)
+    }
+}
+
+impl<C> PgListener<C>
+where
+    C: DerefMut<Target = PgConnection>,
+{
+    /// Get the next async notification from the database.
+    pub async fn next(&mut self) -> Result<NotifyMessage> {
+        loop {
+            match (&mut self.0).receive().await? {
+                Some(Message::NotificationResponse(notification)) => return Ok(notification.into()),
+                Some(msg) => {
+                    return Err(protocol_err!(
+                        "unexpected message received from database {:?}",
+                        msg
+                    )
+                    .into())
+                }
+                None => continue,
+            }
+        }
+    }
+}
+
+impl<C> PgListener<C>
+where
+    C: Connection,
+{
+    /// Close this listener stream and its underlying connection.
+    pub async fn close(self) -> BoxFuture<'static, Result<()>> {
+        self.0.close()
+    }
+}
+
+impl<C> std::ops::Deref for PgListener<C> {
+    type Target = C;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<C> std::ops::DerefMut for PgListener<C> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<C: Connection<Database = Postgres>> crate::Executor for PgListener<C> {
+    type Database = super::Postgres;
+
+    fn send<'e, 'q: 'e>(&'e mut self, query: &'q str) -> BoxFuture<'e, Result<()>> {
+        Box::pin(self.0.send(query))
+    }
+
+    fn execute<'e, 'q: 'e>(
+        &'e mut self,
+        query: &'q str,
+        args: PgArguments,
+    ) -> BoxFuture<'e, Result<u64>> {
+        Box::pin(self.0.execute(query, args))
+    }
+
+    fn fetch<'e, 'q: 'e>(
+        &'e mut self,
+        query: &'q str,
+        args: PgArguments,
+    ) -> BoxStream<'e, Result<PgRow>> {
+        self.0.fetch(query, args)
+    }
+
+    fn describe<'e, 'q: 'e>(
+        &'e mut self,
+        query: &'q str,
+    ) -> BoxFuture<'e, Result<Describe<Self::Database>>> {
+        Box::pin(self.0.describe(query))
+    }
+}
+
+/// An asynchronous message sent from the database.
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct NotifyMessage {
+    /// The channel of the notification, which can be thought of as a topic.
+    pub channel: String,
+    /// The payload of the notification.
+    pub payload: String,
+}
+
+impl From<Box<NotificationResponse>> for NotifyMessage {
+    fn from(src: Box<NotificationResponse>) -> Self {
+        Self {
+            channel: src.channel_name,
+            payload: src.message,
+        }
+    }
+}
+
+/// Build a query which issues a LISTEN command for each given channel.
+fn build_listen_all_query(channels: impl IntoIterator<Item = impl AsRef<str>>) -> String {
+    channels.into_iter().fold(String::new(), |mut acc, chan| {
+        acc.push_str(r#"LISTEN ""#);
+        acc.push_str(chan.as_ref());
```

Comment:
Channels must be a quoted identifier (good) but also must be no more than 63 characters. Should we error here? Postgres just truncates :crying_cat_face:

Reply:
I can't seem to find that in the postgres docs anywhere. Does NOTIFY also truncate the channel? If so, perhaps we just include this information in a doc string and trust the user to do the correct thing? 

Maybe we just do a check and log a warning on the standard logging facade?